File: Components/DevExpressVCL/Demos/ExpressSpreadSheet/CBuilder/Common/ReportDesignerBaseForm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ReportDesignerBaseForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BaseForm"
#pragma link "ReportPreviewUnit"
#pragma link "cxCheckBox"
#pragma link "cxClasses"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxEdit"
#pragma link "cxFilterControl"
#pragma link "cxGraphics"
#pragma link "cxGroupBox"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxSplitter"
#pragma link "dxSpreadSheetCore"
#pragma link "dxSpreadSheetReportDesigner"
#pragma resource "*.dfm"
TfrmReportDesignerBase *frmReportDesignerBase;
//---------------------------------------------------------------------------
__fastcall TfrmReportDesignerBase::TfrmReportDesignerBase(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::AfterConstruction()
{
  TfmBaseForm::AfterConstruction();
  FilterChanged(NULL);
}

//---------------------------------------------------------------------------

TdxSpreadSheet* TfrmReportDesignerBase::GetSpreadSheet()
{
  return (TdxSpreadSheet*)ReportDesigner;
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::FormCreate(TObject *Sender)
{
  Initialize();
  WindowState = wsMaximized;
  ReportDesigner->FieldChooser->Site = cxgFieldChooserSite;
  ReportDesigner->FieldChooser->Show();
  Preview = NULL;
  TfmBaseForm::FormCreate(Sender);
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::FormDestroy(TObject *Sender)
{
  if (FilterFileName != "") {
	Filter->SaveToFile(FilterFileName);
	if (Preview != NULL) Preview->Release();
  }
}


void TfrmReportDesignerBase::Initialize()
{
  // todo: need override
  if (ReportDesigner->DataBinding->DataController->DataSet != NULL ){
	FilterFileName = "..\\..\\Data\\" + ReportDesigner->DataBinding->DataController->DataSet->Name;
	if (FilterFileName != "") {
	  FilterFileName = ChangeFileExt(FilterFileName, ".flt");
	}
  }
  if (FileExists(FilterFileName)) {
	Filter->LoadFromFile(FilterFileName);
	cbxUseFilter->Checked = Filter->FilterText != "";
  }
}

//---------------------------------------------------------------------------

void TfrmReportDesignerBase::LoadDataset(TdxMemData *ADataSet, const UnicodeString AFileName)
{
  TFileStream *AFileStream =  new TFileStream(AFileName, fmOpenRead);
  try {
	ADataSet->DisableControls();
	try {
	  for (int I = ADataSet->Fields->Count - 1; I == 1; I--) {
		if (ADataSet->Fields->Fields[I]->FieldKind == fkData) {
		  ADataSet->Fields->Fields[I]->Free();
		}
	  }
	  ADataSet->CreateFieldsFromStream(AFileStream);
	  AFileStream->Position = 0;
	  ADataSet->LoadFromStream(AFileStream);
	  ADataSet->Active = true;
	}
	__finally {
	  ADataSet->EnableControls();
	}
  }
  __finally {
	AFileStream->Free();
  }
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::miDesignViewClick(TObject *Sender)
{
  ReportDesigner->Options->DesignView = miDesignView->Checked;
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::miPreviewClick(TObject *Sender)
{
  FilterChanged(Sender);
  if (Preview == NULL) Preview = new TfrmPreview(Application);
  Preview->ssResult->ClearAll();
  ReportDesigner->Build(Preview->ssResult);
  Preview->Show();
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::miRemoveClick(TObject *Sender)
{
  int ARow, AColumn;
  TdxSpreadSheetReportSection *ASection;

  ARow = ReportDesigner->ActiveSheetAsTable->Selection->FocusedRow;
  AColumn = ReportDesigner->ActiveSheetAsTable->Selection->FocusedColumn;
  //
  if (ReportDesigner->FindSectionByCell(ARow, AColumn, ASection))
	ReportDesigner->RemoveSection(ASection);
  ReportDesignerSelectionChanged(NULL);
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::OnSectionClick(TObject *Sender)
{
  TRect R;
  R = ReportDesigner->ActiveSheetAsTable->Selection->Area;
  switch (((TComponent*)Sender)->Tag){
	case 0:
	  ReportDesigner->SetHeaderSection(R);
	  return;
	case 1:
	  ReportDesigner->SetDetailSection(R, -1);
	  return;
	case 2:
	  ReportDesigner->SetFooterSection(R);
	  return;
	default: return;
  }
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::FilterChanged(TObject *Sender)
{
  if (cbxUseFilter->Checked) {
	Filter->ApplyFilter();
  }
  ReportDesigner->DataBinding->DataController->Filter->Active = cbxUseFilter->Checked;
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::btnApplyClick(TObject *Sender)
{
  Filter->ApplyFilter();
  cbxUseFilter->Checked = (Filter->FilterText != "");
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::btnClearClick(TObject *Sender)
{
  Filter->Clear();
  cbxUseFilter->Checked = (Filter->FilterText != "");
}

//---------------------------------------------------------------------------

void __fastcall TfrmReportDesignerBase::ReportDesignerSelectionChanged(TObject *Sender)
{
  int ARow, AColumn;
  TdxSpreadSheetReportSection *ASection;

  ARow = ReportDesigner->ActiveSheetAsTable->Selection->FocusedRow;
  AColumn = ReportDesigner->ActiveSheetAsTable->Selection->FocusedColumn;
  //
  if (ReportDesigner->FindSectionByCell(ARow, AColumn, ASection)) {
	miHeader->Checked = (ASection->SectionType == rstHeader);
	miDetail->Checked = (ASection->SectionType == rstDetail);
	miFooter->Checked = (ASection->SectionType == rstFooter);
  }
  else {
	miHeader->Checked = false;
	miDetail->Checked = false;
	miFooter->Checked = false;
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressSpreadSheet/Delphi/Common/ReportDesignerBaseForm.pas
unit ReportDesignerBaseForm;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters, dxSkinsCore, dxSpreadSheetCore,
  dxSpreadSheetFunctions, dxSpreadSheetGraphics, dxSpreadSheetClasses, dxSpreadSheetTypes, SkinDemoUtils,
  dxSpreadSheet, dxSpreadSheetReportDesigner, BaseForm, cxContainer, cxEdit, cxSplitter, cxGroupBox, ExtCtrls, DB,
  DBClient, Menus, cxClasses, ComCtrls, ReportPreviewUnit, cxLabel, cxTextEdit, cxCheckBox, cxFilterControl, dxmdaset,
  StdCtrls;

type
  TfrmReportDesignerBase = class(TfmBaseForm)
    ReportDesigner: TdxSpreadSheetReportDesigner;
    cxSplitter1: TcxSplitter;
    Panel1: TPanel;
    cxSplitter2: TcxSplitter;
    cxgFieldChooserSite: TcxGroupBox;
    cxgFilter: TcxGroupBox;
    miDesignView: TMenuItem;
    miPreview: TMenuItem;
    miSections: TMenuItem;
    N2: TMenuItem;
    asd1: TMenuItem;
    miRemove: TMenuItem;
    miHeader: TMenuItem;
    miFooter: TMenuItem;
    miDetail: TMenuItem;
    N3: TMenuItem;
    cbxUseFilter: TcxCheckBox;
    Filter: TcxFilterControl;
    btnApply: TButton;
    btnClear: TButton;
    procedure FormCreate(Sender: TObject);
    procedure miDesignViewClick(Sender: TObject);
    procedure miPreviewClick(Sender: TObject);
    procedure miRemoveClick(Sender: TObject);
    procedure OnSectionClick(Sender: TObject);
    procedure ReportDesignerSelectionChanged(Sender: TObject);
    procedure FilterChanged(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure btnApplyClick(Sender: TObject);
    procedure btnClearClick(Sender: TObject);
  protected
    FilterFileName: string;
  public
    procedure AfterConstruction; override;
    procedure BeforePreview(ASpreadSheet: TdxSpreadSheet); virtual;
    function GetSpreadSheet: TdxSpreadSheet; override;
    procedure Initialize; virtual;
    procedure LoadDataset(ADataSet: TdxMemData; const AFileName: string);
  end;

var
  frmReportDesignerBase: TfrmReportDesignerBase;

implementation

{$R *.dfm}

procedure TfrmReportDesignerBase.miDesignViewClick(Sender: TObject);
var
  AItem: TdxBaseMenuItem;
begin
  if FindMenuItem('miDesignView', AItem) then
    ReportDesigner.Options.DesignView := GetMenuItemChecked(AItem);
end;

procedure TfrmReportDesignerBase.miPreviewClick(Sender: TObject);
begin
  FilterChanged(nil);
  if Preview = nil then
    Preview := TfrmPreview.Create(nil);
  Preview.ssResult.ClearAll;
  ReportDesigner.Build(Preview.ssResult);
  BeforePreview(Preview.ssResult);
  Preview.Show;
end;

procedure TfrmReportDesignerBase.miRemoveClick(Sender: TObject);
var
  ARow, AColumn: Integer;
  ASection: TdxSpreadSheetReportSection;
begin
  ARow := ReportDesigner.ActiveSheetAsTable.Selection.FocusedRow;
  AColumn := ReportDesigner.ActiveSheetAsTable.Selection.FocusedColumn;
  //
  if ReportDesigner.FindSectionByCell(ARow, AColumn, ASection) then
    ReportDesigner.RemoveSection(ASection);
  ReportDesignerSelectionChanged(nil);
end;

procedure TfrmReportDesignerBase.OnSectionClick(Sender: TObject);
var
  R: TRect;
begin
  R := ReportDesigner.ActiveSheetAsTable.Selection.Area;
  case (Sender as TComponent).Tag of
    0:
      ReportDesigner.SetHeaderSection(R);
    1:
      ReportDesigner.SetDetailSection(R, -1);
    2:
      ReportDesigner.SetFooterSection(R);
  end;
end;

procedure TfrmReportDesignerBase.ReportDesignerSelectionChanged(Sender: TObject);
var
  ARow, AColumn: Integer;
  ASection: TdxSpreadSheetReportSection;
begin
  ARow := ReportDesigner.ActiveSheetAsTable.Selection.FocusedRow;
  AColumn := ReportDesigner.ActiveSheetAsTable.Selection.FocusedColumn;
  //
  if not ReportDesigner.FindSectionByCell(ARow, AColumn, ASection) then
    ASection := nil;
  //
  MenuItemSetChecked('miHeader', (ASection <> nil) and (ASection.SectionType = rstHeader));
  MenuItemSetChecked('miDetail', (ASection <> nil) and (ASection.SectionType = rstDetail));
  MenuItemSetChecked('miFooter', (ASection <> nil) and (ASection.SectionType = rstFooter));
end;

procedure TfrmReportDesignerBase.btnApplyClick(Sender: TObject);
begin
  Filter.ApplyFilter;
  cbxUseFilter.Checked := Filter.FilterText <> '';
end;

procedure TfrmReportDesignerBase.btnClearClick(Sender: TObject);
begin
  Filter.Clear;
  cbxUseFilter.Checked := Filter.FilterText <> '';
end;

procedure TfrmReportDesignerBase.FilterChanged(Sender: TObject);
begin
  if cbxUseFilter.Checked then
    Filter.ApplyFilter;
  ReportDesigner.DataBinding.DataController.Filter.Active := cbxUseFilter.Checked;
end;

procedure TfrmReportDesignerBase.FormCreate(Sender: TObject);
begin
  inherited;
  Initialize;
  WindowState := wsMaximized;
  ReportDesigner.FieldChooser.Site := cxgFieldChooserSite;
  ReportDesigner.FieldChooser.Show;
end;

procedure TfrmReportDesignerBase.FormDestroy(Sender: TObject);
begin
  if FilterFileName <> '' then
  Filter.SaveToFile(FilterFileName);
  FreeAndNil(Preview);
end;

procedure TfrmReportDesignerBase.AfterConstruction;
begin
  inherited AfterConstruction;
  FilterChanged(nil);
end;

function TfrmReportDesignerBase.GetSpreadSheet: TdxSpreadSheet;
begin
  Result := TdxSpreadSheet(ReportDesigner);
end;

procedure TfrmReportDesignerBase.BeforePreview(ASpreadSheet: TdxSpreadSheet);
begin

end;

procedure TfrmReportDesignerBase.Initialize;
begin
  // todo: need override
  if ReportDesigner.DataBinding.DataController.DataSet <> nil then
  begin
    FilterFileName := '..\..\Data\' + ReportDesigner.DataBinding.DataController.DataSet.Name;
    if FilterFileName <> '' then
      FilterFileName := ChangeFileExt(FilterFileName, '.flt');
  end;
  if FileExists(FilterFileName) then
  begin
    Filter.LoadFromFile(FilterFileName);
    cbxUseFilter.Checked := Filter.FilterText <> '';
  end;
end;

procedure TfrmReportDesignerBase.LoadDataset(ADataSet: TdxMemData; const AFileName: string);
var
  I: Integer;
  AFileStream: TFileStream;
begin
  AFileStream := TFileStream.Create(AFileName, fmOpenRead);
  try
    ADataSet.DisableControls;
    try
      for I := ADataSet.Fields.Count - 1 downto 0 do
        if ADataSet.Fields[I].FieldKind = fkData then
          ADataSet.Fields[I].Free;
      ADataSet.CreateFieldsFromStream(AFileStream);
      AFileStream.Position := 0;
      ADataSet.LoadFromStream(AFileStream);
      ADataSet.Active := True;
    finally
      ADataSet.EnableControls;
    end;
  finally
    AFileStream.Free;
  end;
end;

end.

